src/main/java/utility/ImList.java

```diff
@@ -0,0 +1,253 @@
+package utility;
+
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.stream.Stream;
+import java.io.Serializable;
+
+/**
+ * An immutable implementation of the {@code ArrayList} using an 
+ * immutable delegation design pattern.
+ *
+ * @author  cs2030
+ * @param <E> the type of elements in this list
+ */
+public class ImList<E> implements Iterable<E>, Serializable {
```

Comment:
Didn't expect to see familiar stuff from 2030 here :O

Reply:
honestly idk if allowed to use i probably should ask first...